packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,1975 @@
+    });
+
+    return caveatArray && isNonEmptyArray(caveatArray)
+      ? caveatArray
+      : undefined;
+  }
+
+  /**
+   * This methods validates that the specified caveat is an object with the
+   * expected properties and types. It also ensures that a caveat specification
+   * exists for the requested caveat type, and calls the specification
+   * validator, if it exists, on the caveat object.
+   *
+   * Throws an error if validation fails.
+   *
+   * @param caveat - The caveat object to validate.
+   * @param origin - The origin associated with the subject of the parent
+   * permission.
+   * @param target - The target name associated with the parent permission.
+   */
+  private validateCaveat(
+    caveat: unknown,
+    origin: OriginString,
+    target: string,
+  ): void {
+    if (!isPlainObject(caveat)) {
+      throw new InvalidCaveatError(caveat, origin, target);
+    }
+
+    if (Object.keys(caveat).length !== 2) {
+      throw new InvalidCaveatFieldsError(caveat, origin, target);
+    }
+
+    if (typeof caveat.type !== 'string') {
+      throw new InvalidCaveatTypeError(caveat, origin, target);
+    }
+
+    const specification = this.getCaveatSpecification(caveat.type);
+    if (!specification) {
+      throw new UnrecognizedCaveatTypeError(caveat.type, origin, target);
+    }
+
+    if (!hasProperty(caveat, 'value') || caveat.value === undefined) {
+      throw new CaveatMissingValueError(caveat, origin, target);
+    }
+
+    if (!isValidJson(caveat.value)) {
+      throw new CaveatInvalidJsonError(caveat, origin, target);
+    }
+
+    // Typecast: TypeScript still believes that the caveat is a PlainObject.
+    specification.validator?.(caveat as CaveatConstraint, origin, target);
+  }
+
+  /**
+   * Initiates a permission request that requires user approval. This should
+   * always be used to grant additional permissions to a subject, unless user
+   * approval has been obtained through some other means.
+   *
+   * Permissions are validated at every step of the approval process, and this
+   * method will reject if validation fails.
+   *
+   * @see {@link ApprovalController} For the user approval logic.
+   * @see {@link PermissionController.acceptPermissionsRequest} For the method
+   * that _accepts_ the request and resolves the user approval promise.
+   * @see {@link PermissionController.rejectPermissionsRequest} For the method
+   * that _rejects_ the request and the user approval promise.
+   * @param origin - The origin of the grantee subject.
+   * @param requestedPermissions - The requested permissions.
+   * @param id - The id of the permissions request.
+   * @param preserveExistingPermissions - Whether to preserve the subject's
+   * existing permissions.
+   * @returns The granted permissions and request metadata.
+   */
+  async requestPermissions(
+    subject: PermissionSubjectMetadata,
+    requestedPermissions: RequestedPermissions,
+    id = nanoid(),
+    preserveExistingPermissions = true,
```

Comment:
Is this option used anywhere? I couldn't find any cases.

Also, this would be awkward to set if `id` was unset. It opens us up for the easy mistake of setting `id` to `null`, which would break things because the default parameter syntax only applies to `undefined`.

Perhaps we could move `id` and `preserveExistingPermissions` into an "options" object? That would improve readability and avoid the hazard of accidentally setting `id` to null. Unless you think we can remove `preserveExistingPermissions` altogether.

Reply:
I think `preserveExistingPermissions` is worth keeping, but I implemented your "options" suggestion here: e4875ee